internal/cluster.go

```diff
@@ -223,31 +257,13 @@ func (cs *clusterService) logMembers() {
 
 func (cs *clusterService) AddMembershipListener(listener interface{}) string {
 	registrationID, _ := iputil.NewUUID()
-	cs.mu.Lock()
-	defer cs.mu.Unlock()
-	listeners := cs.listeners.Load().(map[string]interface{})
-	copyListeners := make(map[string]interface{}, len(listeners)+1)
-	for k, v := range listeners {
-		copyListeners[k] = v
-	}
-	copyListeners[registrationID] = listener
-	cs.listeners.Store(copyListeners)
+	cs.listeners.Store(registrationID, listener)
 	return registrationID
 }
 
 func (cs *clusterService) RemoveMembershipListener(registrationID string) bool {
-	cs.mu.Lock()
-	defer cs.mu.Unlock()
-	listeners := cs.listeners.Load().(map[string]interface{})
-	copyListeners := make(map[string]interface{}, len(listeners)-1)
-	for k, v := range listeners {
-		copyListeners[k] = v
-	}
-	_, found := copyListeners[registrationID]
-	if found {
-		delete(copyListeners, registrationID)
-	}
-	cs.listeners.Store(copyListeners)
+	_, found := cs.listeners.Load(registrationID)
```

Comment:
this change is not preserving behaviour. Two concurrent call to `RemoveMembershipListener` can return true, eventough only one of them makes the actual delete.  I have checked the sync.Map interface but could not find a fitting solution on it.

Reply:
I have used a mutex instead of defining a new data structure that extends sync.Map as this probably wont be needed at different places of the codebase. If we need to use this a lot in the future, we can define a new data structure this way.